data_handler/handlers/loan_states/hashtack_v0/interest_rate.py

```diff
@@ -1,87 +1,168 @@
+            logging.info("Invalid block numbers provided.")
+            self.events.clear()
+        self.events.clear()
         result = self.api_connector.get_data(
             HASHSTACK_INTEREST_RATE_ADDRESS,
-            self.min_block_number,
-            self.max_block_number,
+            start_block,
+            end_block,
         )
         if isinstance(result, dict):
-            raise RuntimeError(f"Error while fetching events: {result.get('error', 'Unknown error')}")
-        self.events = list(filter(lambda event: event["data"][0] == self.token, result))
+            logging.info(f"Error while fetching events: {result.get('error', 'Unknown error')}")
+            self.events = []
+            self._events_over = True
+            return
+        self.events = result
 
-    def calculate_interest_rates(self) -> None:
-        """Calculate the interest rates for the given token."""
+    def _add_block_data(self, block_data: list[InterestRate], interest_rate_entry: InterestRate) -> None:
+        """
+        Add the interest rate entry to the block data and update last block data stored.
+        :param block_data: list[InterestRate] - list of blocks with interest rates data.
+        :param interest_rate_entry: InterestRate - interest rate entry to add.
+        """
+        block_data.append(interest_rate_entry)
+        self.last_block_data = interest_rate_entry
+
+    def calculate_interest_rates(self) -> list[InterestRate]:
+        """
+        Calculate the interest rates for provided events range.
+        :return: list[InterestRate] - list of blocks with interest rates data.
+        """
         if not self.events:
-            return
-        supply_interest_rate_cum = Decimal("0")
-        borrow_interest_rate_cum = Decimal("0")
-        percents_decimals_shift = Decimal(10000)
+            return []
+        percents_decimals_shift = Decimal("0.0001")
+        blocks_data: list[InterestRate] = []
+        interest_rate_state = InterestRateState(self.events[0]["block_number"], self.last_block_data)
+
         for index, event in enumerate(self.events):
-            # First block will have 0 interest rate
-            if index == 0:
-                first_event = HashstackV0Event(
-                    event["block_number"], event["timestamp"], supply_interest_rate_cum, borrow_interest_rate_cum,
-                )
-                self.interest_rates.append(first_event)
+            # If block number in event is different from previous, add block data
+            if interest_rate_state.current_block != event["block_number"]:
+                self._add_block_data(blocks_data, interest_rate_state.build_interest_rate_model(HASHSTACK_ID))
+
+            # Get token name. Validate event `key_name` and token name.
+            token_name = TOKEN_MAPPING.get(event["data"][0], "")
+            if not token_name or event["key_name"] != "current_apr":
+                interest_rate_state.current_block = event["block_number"]
```

Comment:
Why are you updating just the block number and not the timestamp? They seem to be qualitatively similar variables.

Reply:
Also, it seems that the block number and timestam could just be updated with every event.